hudi-spark-datasource/hudi-spark3.3.x/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala

```diff
@@ -223,6 +215,20 @@ private[sql] class AvroSerializer(
         val numFields = st.length
         (getter, ordinal) => structConverter(getter.getStruct(ordinal, numFields))
 
+      ////////////////////////////////////////////////////////////////////////////////////////////
+      // Following section is amended to the original (Spark's) implementation
+      // >>> BEGINS
+      ////////////////////////////////////////////////////////////////////////////////////////////
+
+      case (st: StructType, UNION) =>
```

Comment:
Folks, we need to be careful when we add support for Spark 3.3 we should review the changes against previous versions b/c some of the code we re-home from Spark has actually our own fixes which we can't upstream to Spark in a timely fashion.

This change in particular if missed would have made Column Stats Index unreadable on Spark 3.3. 
I had added annotations like you can see here to make it more catchy to the eye that this code isn't just a verbatim copy from Spark, but we need to be careful when we try to back-port the changes. 

I'd recommend following procedure: 

1. By default, carry over implementation from previous Spark version.
2. Then reconcile it w/ the latest Spark version (back-porting)


cc @CTTY @rahil-c @umehrot2 @nsivabalan @yihua @codope @xushiyan

Reply:
Very good point! Sounds good. Let's make sure that the annotation is common across code, easily searchable. Probably, we should put this in coding guidelines as well? https://hudi.apache.org/contribute/developer-setup#coding-guidelines